Fixed the way json errors are requested from cargo & rustc

White-Oak opened the pull request:
Closes #69

Also fixed the way envinronment variables were passed to BufferedProcess --
according to [the documentation](https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback) (*`BufferedProcess` is actually a wrapper above node's exec*) 
they should be passed in an `env` subobject of `options`, but they were passed 'as is' in the options object.
Adding to these environment variables accordingly set `RUSTFLAGS` variable, when JSON errors are available, makes cargo tell compiler to output JSON errors.
When using `rustc` simply providing `--error-format=json` is enough to make compiler do what we want.

White-Oak commented:
Ah, I've also added the code to always redirect errors to Atom from `stderr` when Atom is run in dev mode.

Arcanemagus commented on lib/linter-rust.coffee:
```diff
@@ -21,8 +21,9 @@ class LinterRust
       PATH = path.dirname @cmd[0]
-      options = JSON.parse JSON.stringify process.env
-      options.PATH = PATH + path.delimiter + options.PATH
+      options = new Object
```
Any reason you didn't just use `{}`?

White-Oak replied:
@Arcanemagus the only reason for this is my poor knowledge of CoffeeScript :) I'll fix this.

Arcanemagus replied:
Yea, every time I see linters still stuck on CoffeeScript I cringe, I'll get to moving them to ES2015 eventually! 😛.

Arcanemagus replied:
http://decaffeinate-project.org/repl/

^ This is great for getting readable code out of the CoffeeScript

Arcanemagus commented on lib/linter-rust.coffee:
```diff
@@ -49,6 +55,9 @@ class LinterRust
         else
           resolve []
 
+      if do @ableToJSONErrors
```
Since this is now being called _three_ times for every lint call, it's results should probably be cached...

White-Oak replied:
@Arcanemagus I'll cache it for a single linter usage (meaning per-`lint` call), but I think it would be incorrect to cache it globally (I mean for a duration of Atom running) -- a user may switch between versions of rustc, while still using the same Atom process running).

Arcanemagus commented on lib/linter-rust.coffee:
```diff
@@ -34,6 +35,11 @@ class LinterRust
             dismissable: true
+        else
+          if do atom.inDevMode
+            atom.notifications.addError "Something wrong",
```
"Something went wrong" (running rustc?)

White-Oak replied:
@Arcanemagus is it a proposal for a better error message? If yes, that sounds good, I'll fix it.

Arcanemagus commented:
That's all I see just looking at the code.

White-Oak commented:
@Arcanemagus I've added a cache for 'ability to have JSON errors', fixed 'new object' stuff and the way dev mode notifications look:
![dev mode notifications](https://cloud.githubusercontent.com/assets/2813216/18219028/3f0a5818-716f-11e6-96aa-9664ac129908.png)

It is possible to stack them in one error message, if one would prefer it that way.

Arcanemagus commented:
As it's already behind a dev mode check, it's probably fine to leave them like that.

gmist commented on lib/linter-rust.coffee:
```diff
@@ -21,11 +22,14 @@ class LinterRust
       options.cwd = curDir
       command = @cmd[0]
       args = @cmd.slice 1
+      @cachedAbleToJsonErrors = null
```
hmm... is this line really necessary?

White-Oak replied:
@gmist This ensures that the cached value is live only for a one lint process.

gmist replied:
Ah, exactly
